src/main/java/io/cdap/plugin/gcp/gcs/GCSPath.java

```diff
@@ -83,28 +86,31 @@ public int hashCode() {
    * @throws IllegalArgumentException if the path string is invalid
    */
   public static GCSPath from(String path) {
-    URI uri = getURI(path);
-    String bucket = uri.getAuthority();
-    String name = uri.getPath();
-    // strip preceding '/'. An empty name means it's for a bucket.
-    name = name.isEmpty() ? name : name.substring(1);
-    return new GCSPath(uri, bucket, name);
-  }
+    if (path.isEmpty()) {
+      throw new IllegalArgumentException("GCS path can not be empty. The path must be of form " +
+                                           "'gs://<bucket-name>/path'.");
+    }
+
+    if (path.startsWith(ROOT_DIR)) {
+      path = path.substring(1);
+    } else if (path.startsWith(SCHEME)) {
+      path = path.substring(SCHEME.length());
+    }
+
+    String bucket = path;
+    int idx = path.indexOf(ROOT_DIR);
+    // if the path within bucket is provided, then only get the bucket
+    if (idx > 0) {
+      bucket = path.substring(0, idx);
+    }
 
-  private static URI getURI(String path) {
-    try {
-      URI uri = new URI(path);
-      if (uri.getScheme() != null && (!uri.getScheme().equalsIgnoreCase("gs") || uri.getAuthority() == null)) {
-        throw new IllegalArgumentException(String.format("Invalid path '%s'. The path must be of form " +
-                                                           "'gs://<bucket-name>/path'.", path));
-      }
-      if (uri.getScheme() == null) {
-        return path.startsWith(ROOT_DIR) ? new URI("gs:/" + path) : new URI("gs://" + path);
-      }
-      return uri;
-    } catch (URISyntaxException e) {
-      throw new IllegalArgumentException(String.format("Invalid path '%s'. The path must be of form " +
-                                                         "'gs://<bucket-name>/path'.", path), e);
+    if (!Pattern.matches("[a-zA-Z0-9-_.]+", bucket)) {
+      throw new IllegalArgumentException(String.format("Invalid bucket name in path '%s'. Bucket name should only " +
+                                                         "contain alphanumeric, '-'. '_' and '.'.", path));
     }
+
+    String file = idx > 0 ? path.substring(idx) : "";
+    URI uri = URI.create(SCHEME + bucket + UrlEscapers.urlFragmentEscaper().escape(file));
+    return new GCSPath(uri, bucket, file.replaceAll("^/", ""));
```

Comment:
Why need to do the `replaceAll` with regex? If the file is `"/abc/def"`, then it will become `"abcdec"`.. is it desired? It won't match with what's in the URI. I think ideally the `GCPath` constructor should only takes `URI`, as both `bucket` and `file` should be derived from that.

Reply:
This regex only replaces leading `/` because current behavior expects name without leading `/`.
![image](https://user-images.githubusercontent.com/14131070/57483565-f59fed00-725b-11e9-9280-c67311c993dd.png)